crates/tabby-download/src/lib.rs

```diff
@@ -41,49 +42,98 @@ async fn download_model_impl(
                 registry.name, name
             );
-            fs::remove_file(&model_path)?;
+
+            fs::remove_dir_all(registry.get_model_store_dir(name))?;
         } else {
             return Ok(());
         }
     }
 
-    let Some(model_url) = model_info
-        .urls
-        .iter()
-        .flatten()
-        .find(|x| select_by_download_host(x))
-    else {
-        return Err(anyhow!("No valid url for model <{}>", model_info.name));
-    };
-
-    // Replace the huggingface.co domain with the mirror host if it is set.
-    let model_url = if let Ok(host) = std::env::var("TABBY_HUGGINGFACE_HOST_OVERRIDE") {
-        model_url.replace("huggingface.co", &host)
-    } else {
-        model_url.to_owned()
-    };
-
-    let strategy = ExponentialBackoff::from_millis(100).map(jitter).take(2);
-    let download_job = Retry::spawn(strategy, || {
-        download_file(&model_url, model_path.as_path(), &model_info.sha256)
-    });
-    download_job.await?;
-    Ok(())
-}
+    let urls = model_info.filter_download_address();
+
+    if urls.is_empty() {
+        bail!(
+            "No download URLs available for <{}/{}>",
+            registry.name,
+            model_info.name
+        );
+    }
 
-async fn download_file(url: &str, path: &Path, expected_sha256: &str) -> Result<()> {
-    let dir = path
-        .parent()
-        .ok_or_else(|| anyhow!("Must not be in root directory"))?;
+    // prepare for download
+    let dir = registry.get_model_store_dir(name);
     fs::create_dir_all(dir)?;
 
-    let filename = path
-        .to_str()
-        .ok_or_else(|| anyhow!("Could not convert filename to UTF-8"))?;
-    let intermediate_filename = filename.to_owned() + ".tmp";
+    let mut download_tasks = vec![];
+    for (index, url) in urls.iter().enumerate() {
+        let dir = registry
+            .get_model_store_dir(name)
+            .to_string_lossy()
+            .into_owned();
+        let filename = format!("{:05}-of-{:05}", index, urls.len());
+        let strategy = ExponentialBackoff::from_millis(100).map(jitter).take(2);
+        download_tasks.push(Retry::spawn(strategy, move || {
+            let dir = dir.clone();
+            let filename = filename.clone();
+            download_file(&url.0, dir, filename, &url.1)
+        }));
+    }
 
-    let mut bar = WrappedBar::new(0, url, false);
+    let results = join_all(download_tasks).await;
+
+    // handle errors
+    let errors: Vec<anyhow::Error> = results
+        .into_iter()
+        .filter_map(|e| if let Err(e) = e { Some(e) } else { None })
+        .collect();
+    if errors.is_empty() {
+        Ok(())
+    } else {
+        let combined_error = errors
+            .into_iter()
+            .fold(anyhow::anyhow!("Multiple errors occurred"), |acc, err| {
+                acc.context(err)
+            });
+        Err(combined_error)
+    }
+}
 
+async fn tryget_download_filename(url: &str) -> Result<String> {
```

Comment:
since we’re storing model in format of `model-000xxx-of-000yyy.gguf`, do we still need this function to extract file name?